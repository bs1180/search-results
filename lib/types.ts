export enum Currency {
  EUR = "EUR",
  AUD = "AUD",
  USD = "USD"
}

export interface RawSearchResult {
  id: number;
  tour_name: string;
  length: number;
  description: string;
  price: number;
  saving: number;
  currency: Currency;
  destinations: string[];
  age_from: number;
  age_to: number;
  rating: number;
  tour_operator: string;
  country: string;
  tour_image: string;
  map_image: string;
}

export interface SearchResult extends RawSearchResult {
  price_eur: number;
}

export enum SortOrder {
  lowestPrice = "lowest-price",
  highestPrice = "highest-price",
  shortestTour = "shortest-tour",
  longestTour = "longest-tour"
}

export type QueryParams = {
  sort: SortOrder;
  page: string;
};
