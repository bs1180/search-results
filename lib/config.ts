export const API_URL = "https://api.myjson.com/bins/18x6yt";
export const EXCHANGE_RATES_URL = "https://api.exchangeratesapi.io/latest?symbols=USD,AUD&base=EUR";
export const RESULTS_SIZE = 50;
