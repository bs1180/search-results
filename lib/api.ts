import "isomorphic-fetch";
import { SortOrder, SearchResult } from "./types";
import { EXCHANGE_RATES_URL, API_URL, RESULTS_SIZE } from "./config";

export const getExchangeRates = (): Promise<{ AUD: number; USD: number }> =>
  fetch(EXCHANGE_RATES_URL)
    .then(res => res.json())
    .then(res => res.rates);

const getFakeResults = () => fetch(API_URL).then(res => res.json());

const getSorterFn = (sort: SortOrder): ((a, b) => number) => {
  switch (sort) {
    case SortOrder.highestPrice:
      return (b, a) => a.price_eur - b.price_eur;
    case SortOrder.lowestPrice:
      return (a, b) => a.price_eur - b.price_eur;
    case SortOrder.longestTour:
      return (b, a) => a.length - b.length;
    case SortOrder.shortestTour:
      return (a, b) => a.length - b.length;
  }
};

export const getSearchResults = async ({
  sort,
  page = 1
}: {
  sort?: SortOrder;
  page: number;
}): Promise<{ results: SearchResult[]; count: number; sortedBy: SortOrder }> => {
  // Search results are in multiple currencies, so we do a lookup of the current exchange rates and convert AUD and USD into EUR

  const [exchangeRates, searchResults] = await Promise.all([getExchangeRates(), getFakeResults()]);

  const multiplier = {
    AUD: exchangeRates.AUD,
    USD: exchangeRates.USD,
    EUR: 1
  };

  let fullResults = searchResults.map(result => ({
    price_eur: parseFloat((result.price * multiplier[result.currency]).toFixed(0)),
    ...result
  }));

  // The fake API doesn't support sorting or pagination so we'll do it here
  if (sort) {
    fullResults = fullResults.sort(getSorterFn(sort));
  }

  const begin = (page - 1) * RESULTS_SIZE;
  const end = begin + RESULTS_SIZE;

  return {
    results: fullResults.slice(begin, end),
    count: fullResults.length,
    sortedBy: sort
  };
};
