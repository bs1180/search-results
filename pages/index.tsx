import * as React from "react";
import { NextContext } from "next";
import { withRouter, WithRouterProps } from "next/router";
import { getSearchResults } from "../lib/api";
import { SearchResult, SortOrder, QueryParams } from "../lib/types";
import Card from "../components/Card";
import Select from "../components/Select";
import EmptyResults from "../components/EmptyResults";
import Layout from "../components/Layout";
import Paginator from "../components/Paginator";
import { RESULTS_SIZE } from "../lib/config";
if (process.env.NODE_ENV !== "test") {
  require("../styles/styles.css");
}

interface Props {
  results: SearchResult[];
  count: number;
  sortedBy: SortOrder;
}

export class SearchResultsPage extends React.Component<WithRouterProps<QueryParams> & Props> {
  static defaultProps = {
    results: []
  };

  static async getInitialProps(ctx: NextContext<QueryParams>) {
    const { sort, page } = ctx.query;

    const { results, count, sortedBy } = await getSearchResults({ sort, page: page ? parseInt(page) : undefined });

    return { results, count, sortedBy };
  }

  handleSort = (e: React.ChangeEvent<HTMLSelectElement>) => {
    // We put search parameters (i.e. sort order) in the URL to give a more consistent user experience if they bookmark it, share it with others etc.
    // Changing the sort order also returns to the first page of results
    const url = {
      pathname: "/",
      query: {
        sort: e.target.value
      }
    };
    this.props.router.push(url, url);
  };

  render() {
    const {
      results,
      count,
      router: {
        query: { sort }
      }
    } = this.props;

    const sortable = [
      { value: SortOrder.highestPrice, label: "Highest Price" },
      { value: SortOrder.lowestPrice, label: "Lowest Price" },
      { value: SortOrder.longestTour, label: "Longest Tour" },
      { value: SortOrder.shortestTour, label: "Shortest Tour" }
    ];

    const totalPages = Math.ceil(count / RESULTS_SIZE);

    return (
      <Layout title="Search Results">
        {results.length === 0 ? (
          <EmptyResults />
        ) : (
          <div className="max-w-xl mx-auto flex flex-col my-4 px-4">
            <div className="flex justify-between items-center flex-col-reverse sm:flex-row">
              <Paginator pageCount={totalPages} />
              <Select onChange={this.handleSort} value={sort} options={sortable} />
            </div>
            {results.map(result => (
              <Card key={result.id} result={result} />
            ))}
            <Paginator pageCount={totalPages} />
          </div>
        )}
      </Layout>
    );
  }
}

export default withRouter(SearchResultsPage);
