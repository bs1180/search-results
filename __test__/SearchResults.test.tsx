import * as React from "react";
import { render } from "react-testing-library";
import { SearchResultsPage } from "../pages";
import * as api from "../lib/api";
jest.mock("../lib/api");

describe("Homepage", () => {
  it("calls API from getInitialProps correctly", async () => {
    jest.spyOn(api, "getSearchResults").mockImplementation(() => Promise.resolve({ results: [], count: 0 }));

    const mockContext: any = {
      query: {}
    };

    const res = await SearchResultsPage.getInitialProps(mockContext);

    expect(res.count).toEqual(0);
  });
  it("passes querystring values to API", async () => {
    const mockApi = jest
      .spyOn(api, "getSearchResults")
      .mockImplementation(() => Promise.resolve({ results: [], count: 0 }));

    const mockContext: any = {
      query: {
        page: 2
      }
    };

    await SearchResultsPage.getInitialProps(mockContext);

    expect(mockApi).toHaveBeenCalledWith({ page: 2 });
  });
  it("renders empty results if no results", async () => {
    const props: any = {
      results: [],
      router: {
        query: {}
      }
    };

    const { getByText } = render(<SearchResultsPage {...props} />);

    expect(getByText("Sorry, no results")).toMatchSnapshot();
  });
});
