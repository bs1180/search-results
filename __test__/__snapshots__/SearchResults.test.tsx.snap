// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Homepage renders empty results if no results 1`] = `
<div
  class="h-64 flex items-center justify-center"
>
  Sorry, no results
</div>
`;
